packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,401 @@
+                            // so don't do anything to just continue the inner loop
+                            Freshness::Fresh => (),
+                            Freshness::Stale | Freshness::Resolved => {
+                                if !found_stale_or_resolved_dependency {
+                                    freshen_stack.push(state_var_ptr);
+                                    found_stale_or_resolved_dependency = true;
+                                }
+
+                                // Recurse by putting two items on the stack
+                                // and then continuing at start of stack loop
+                                freshen_stack.push(new_state_var_ptr);
+                            }
+                            Freshness::Unresolved => {
+                                panic!("How did a stale state variable depend on an unresolved state variable?")
+                            }
+                        };
+                    }
+                    // nothing to add to stack if have an essential state variable
+                    // as they cannot be stale
+                    DependencySource::Essential { .. } => (),
+                }
+            }
+        }
+
+        if !found_stale_or_resolved_dependency {
+            // All the dependencies of state_var_ptr are fresh.
+            // We calculate its value if a dependency has changed,
+            // else restore its previous value.
+
+            let component_idx = state_var_ptr.component_idx;
+            let state_var_idx = state_var_ptr.state_var_idx;
+
+            let mut comp = components[component_idx].borrow_mut();
+            let state_var = &mut comp.get_state_variables_mut()[state_var_idx];
+
+            // Check if any dependency has changed since we last called record_all_dependencies_viewed.
+            if state_var.check_if_any_dependency_changed_since_last_viewed() {
+                state_var.calculate_state_var_from_dependencies_and_mark_fresh();
+                state_var.record_all_dependencies_viewed();
+            } else {
+                state_var.restore_previous_value();
+            }
+        }
+    }
+}
+
+#[tailcall]
+pub fn resolve_state_var(
+    state_var_ptr: StateVarPointer,
+    components: &Vec<Rc<RefCell<ComponentEnum>>>,
+    dependencies: &mut Vec<Vec<Vec<Vec<Dependency>>>>,
+    dependent_on_state_var: &mut Vec<Vec<Vec<StateVarPointer>>>,
+    dependent_on_essential: &mut Vec<HashMap<EssentialDataOrigin, Vec<StateVarPointer>>>,
+    essential_data: &mut Vec<HashMap<EssentialDataOrigin, EssentialStateVar>>,
+    // freshen_stack: &mut Vec<StateVarCalculationState>,
+    should_initialize_essential_data: bool,
+) {
+    let component_idx = state_var_ptr.component_idx;
+    let state_var_idx = state_var_ptr.state_var_idx;
+
+    let dependency_instructions;
+
+    {
+        let component = components[component_idx].borrow();
+        let state_var = &component.get_state_variables()[state_var_idx];
```

Comment:
Can you explicitly type this. rust-analyzer is having trouble...

Reply:
Something is wrong with your rust-analyzer! All of these types come directly from the assignment...